src/typing/type-inferencer.ts

```diff
@@ -0,0 +1,35 @@
+import { DefaultNormalizedAstFold } from '../abstract-interpretation/normalized-ast-fold';
+import type { RLogical } from '../r-bridge/lang-4.x/ast/model/nodes/r-logical';
+import type { RNumber } from '../r-bridge/lang-4.x/ast/model/nodes/r-number';
+import type { RString } from '../r-bridge/lang-4.x/ast/model/nodes/r-string';
+
+export enum RDataType {
+    Logical = 'LogicalType',
+    Numeric = 'NumericType',
+    String = 'StringType',
+}
+
+export type TypingInfo = RDataType | undefined;
+
+export class TypeInferencer<Info> extends DefaultNormalizedAstFold<TypingInfo, Info> {
+	constructor() {
+		super(undefined);
+	}
+
+	protected override concat(a: TypingInfo, b: TypingInfo): TypingInfo {
+		// Do not combine typing information for now
+		return b;
```

Comment:
I think to stay correct this has to yield TOP. Because otherwise `1 + TRUE` should yield `logical`, and even function calls may just "pass along" the type.

Reply:
I think it is a good idea to add 2-3 more tests for cases like `1 + TRUE` just to make sure we do not wrongly return a specific type.